In [24]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
%matplotlib inline

In [25]:
credit = pd.read_csv("C:\\Users\\Nik\\Downloads\\creditcard.csv")

In [26]:
credit.head(10)

Unnamed: 0 card  reports       age  income     share  expenditure owner  \
0           1  yes        0  37.66667  4.5200  0.033270   124.983300   yes   
1           2  yes        0  33.25000  2.4200  0.005217     9.854167    no   
2           3  yes        0  33.66667  4.5000  0.004156    15.000000   yes   
3           4  yes        0  30.50000  2.5400  0.065214   137.869200    no   
4           5  yes        0  32.16667  9.7867  0.067051   546.503300   yes   
5           6  yes        0  23.25000  2.5000  0.044438    91.996670    no   
6           7  yes        0  27.91667  3.9600  0.012576    40.833330    no   
7           8  yes        0  29.16667  2.3700  0.076434   150.790000   yes   
8           9  yes        0  37.00000  3.8000  0.245628   777.821700   yes   
9          10  yes        0  28.41667  3.2000  0.019780    52.580000    no   

  selfemp  dependents  months  majorcards  active  
0      no           3      54           1      12  
1      no           3      34           1      13  
2      no           4      58           1       5  
3      no           0      25           1       7  
4      no           2      64           1       5  
5      no           0      54           1       1  
6      no           2       7           1       5  
7      no           0      77           1       3  
8      no           0      97           1       6  
9      no           0      65           1      18

In [27]:
credit.isnull().sum()

Unnamed: 0     0
card           0
reports        0
age            0
income         0
share          0
expenditure    0
owner          0
selfemp        0
dependents     0
months         0
majorcards     0
active         0
dtype: int64

In [28]:
credit.drop(labels='Unnamed: 0', axis=1,inplace=True)

In [29]:
credit.columns

Index(['card', 'reports', 'age', 'income', 'share', 'expenditure', 'owner',
       'selfemp', 'dependents', 'months', 'majorcards', 'active'],
      dtype='object')

In [30]:
credit.drop(labels=['age','income','expenditure','owner','selfemp','dependents','months','majorcards','active'], axis=1,inplace=True)

In [31]:
credit

card  reports     share
0     yes        0  0.033270
1     yes        0  0.005217
2     yes        0  0.004156
3     yes        0  0.065214
4     yes        0  0.067051
...   ...      ...       ...
1314  yes        0  0.002146
1315   no        5  0.000376
1316  yes        0  0.026513
1317  yes        0  0.008999
1318  yes        0  0.111619

[1319 rows x 3 columns]

In [32]:
credit['card']=credit['card'].map({'yes':1,'no':0})

In [33]:
credit

card  reports     share
0        1        0  0.033270
1        1        0  0.005217
2        1        0  0.004156
3        1        0  0.065214
4        1        0  0.067051
...    ...      ...       ...
1314     1        0  0.002146
1315     0        5  0.000376
1316     1        0  0.026513
1317     1        0  0.008999
1318     1        0  0.111619

[1319 rows x 3 columns]

In [34]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

In [17]:
# Let's divide into train and test set

X_train, X_test, y_train, y_test = train_test_split(
    credit.drop(labels='card', axis=1),  # predictors
    credit['card'],  # target
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((923, 2), (396, 2))

In [35]:
# the scaler - for standardisation
from sklearn.preprocessing import StandardScaler

In [36]:
# standardisation: with the StandardScaler from sklearn

# set up the scaler
scaler = StandardScaler()
# fit the scaler to the train set, it will learn the parameters
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [37]:
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

### performing logistic reggression¶

In [38]:
def run_logistic(X_train, X_test, y_train, y_test):

    # function to train and test the performance of logistic regression
    logit = LogisticRegression(random_state=44, C=0.01)
    logit.fit(X_train, y_train)

    print('Train set')
    pred = logit.predict_proba(X_train)
    print(
        'Logistic Regression roc-auc: {}'.format(roc_auc_score(y_train, pred[:, 1])))

    print('Test set')
    pred = logit.predict_proba(X_test)
    print(
        'Logistic Regression roc-auc: {}'.format(roc_auc_score(y_test, pred[:, 1])))

In [39]:
run_logistic(X_train_scaled, X_test_scaled, y_train, y_test)

Train set
Logistic Regression roc-auc: 0.9655392448588243
Test set
Logistic Regression roc-auc: 0.9557719696678085


C:\Users\Nik\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


## Train set
Logistic Regression roc-auc: 0.9655392448588243

Test set

Logistic Regression roc-auc: 0.9557719696678085

# if i select more then 3 columns #lets again perform logistic regrssion on same data with few more columns


In [113]:
credit1 = pd.read_csv("C:\\Users\\Nik\\Downloads\\creditcard.csv")

In [114]:
credit1.drop(labels='Unnamed: 0', axis=1,inplace=True)

In [115]:
credit1['card']=credit1['card'].map({'yes':1,'no':0})

In [116]:
credit1['owner']=credit1['owner'].map({'yes':1,'no':0})

In [117]:
credit1['selfemp']=credit1['selfemp'].map({'yes':1,'no':0})

In [118]:
credit1

card  reports       age  income     share  expenditure  owner  selfemp  \
0        1        0  37.66667  4.5200  0.033270   124.983300      1        0   
1        1        0  33.25000  2.4200  0.005217     9.854167      0        0   
2        1        0  33.66667  4.5000  0.004156    15.000000      1        0   
3        1        0  30.50000  2.5400  0.065214   137.869200      0        0   
4        1        0  32.16667  9.7867  0.067051   546.503300      1        0   
...    ...      ...       ...     ...       ...          ...    ...      ...   
1314     1        0  33.58333  4.5660  0.002146     7.333333      1        0   
1315     0        5  23.91667  3.1920  0.000376     0.000000      0        0   
1316     1        0  40.58333  4.6000  0.026513   101.298300      1        0   
1317     1        0  32.83333  3.7000  0.008999    26.996670      0        1   
1318     1        0  48.25000  3.7000  0.111619   344.157500      1        0   

      dependents  months  majorcards  active  
0              3      54           1      12  
1              3      34           1      13  
2              4      58           1       5  
3              0      25           1       7  
4              2      64           1       5  
...          ...     ...         ...     ...  
1314           0      94           1      19  
1315           3      12           1       5  
1316           2       1           1       2  
1317           0      60           1       7  
1318           2       2           1       0  

[1319 rows x 12 columns]

In [119]:
credit1.columns

Index(['card', 'reports', 'age', 'income', 'share', 'expenditure', 'owner',
       'selfemp', 'dependents', 'months', 'majorcards', 'active'],
      dtype='object')

In [120]:
credit1.drop(labels=['expenditure','owner','selfemp','dependents','months','majorcards'], axis=1,inplace=True)

In [121]:
credit1

card  reports       age  income     share  active
0        1        0  37.66667  4.5200  0.033270      12
1        1        0  33.25000  2.4200  0.005217      13
2        1        0  33.66667  4.5000  0.004156       5
3        1        0  30.50000  2.5400  0.065214       7
4        1        0  32.16667  9.7867  0.067051       5
...    ...      ...       ...     ...       ...     ...
1314     1        0  33.58333  4.5660  0.002146      19
1315     0        5  23.91667  3.1920  0.000376       5
1316     1        0  40.58333  4.6000  0.026513       2
1317     1        0  32.83333  3.7000  0.008999       7
1318     1        0  48.25000  3.7000  0.111619       0

[1319 rows x 6 columns]

In [122]:
#some of the encoder can not able to create realtion ship with categorical type target , so here i m using custom fuction to make the 
#the target as integer 
#target is labeled as integer then no need to do this transformation 

def find_category_mappings(df, variable):
    return {k: i for i, k in enumerate(df[variable].unique(), 0)}

def integer_encode(data,  variable, ordinal_mapping):

    data[variable] = data[variable].map(ordinal_mapping)

In [123]:
for variable in ['card']:
    mappings = find_category_mappings(credit1, variable)
    integer_encode(credit1, variable, mappings)

In [130]:
# Let's divide into train and test set

X_train, X_test, y_train, y_test = train_test_split(
    credit1.drop(labels='card', axis=1),  # predictors
    credit1['card'],  # target
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((923, 5), (396, 5))

In [131]:
# the scaler - for standardisation
from sklearn.preprocessing import StandardScaler

In [132]:
# standardisation: with the StandardScaler from sklearn

# set up the scaler
scaler = StandardScaler()
# fit the scaler to the train set, it will learn the parameters
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [133]:
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# performing logistic reggression

In [134]:
def run_logistic(X_train, X_test, y_train, y_test):

    # function to train and test the performance of logistic regression
    logit = LogisticRegression(random_state=44, C=0.01)
    logit.fit(X_train, y_train)

    print('Train set')
    pred = logit.predict_proba(X_train)
    print(
        'Logistic Regression roc-auc: {}'.format(roc_auc_score(y_train, pred[:, 1])))

    print('Test set')
    pred = logit.predict_proba(X_test)
    print(
        'Logistic Regression roc-auc: {}'.format(roc_auc_score(y_test, pred[:, 1])))

In [135]:
run_logistic(X_train_scaled, X_test_scaled, y_train, y_test)

Train set
Logistic Regression roc-auc: 0.9518415658268862
Test set
Logistic Regression roc-auc: 0.9299434158358674


C:\Users\Nik\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


### Train set
Logistic Regression roc-auc: 0.9518415658268862

Test set

Logistic Regression roc-auc: 0.9299434158358674

#### when we take more columns or data points roc value may decrease 